standalone/pruntime/enclave/src/contracts/mod.rs

```diff
@@ -391,3 +186,212 @@ impl<'de> Visitor<'de> for AcidVisitor {
+        QReq: Serialize + DeserializeOwned + Debug,
+        QResp: Serialize + DeserializeOwned + Debug,
+        Con: NativeContract<Cmd = Cmd, Event = Event, QReq = QReq, QResp = QResp>,
+        PushCommand<Cmd>: Decode + BindTopic + Debug,
+    {
+        pub fn new(
+            contract: Con,
+            send_mq: MessageChannel,
+            cmd_rcv_mq: PeelingReceiver<PushCommand<Cmd>, CmdWrp, CmdPlr>,
+            event_rcv_mq: PeelingReceiver<Event, EventWrp, EventPlr>,
+            ecdh_key: KeyPair,
+        ) -> Self {
+            NativeCompatContract {
+                contract,
+                send_mq,
+                cmd_rcv_mq,
+                event_rcv_mq,
+                ecdh_key,
+            }
+        }
+    }
+
+    impl<Con, Cmd, CmdWrp, CmdPlr, Event, EventWrp, EventPlr, QReq, QResp> Contract
+        for NativeCompatContract<Con, Cmd, CmdWrp, CmdPlr, Event, EventWrp, EventPlr, QReq, QResp>
+    where
+        Cmd: Decode + BindTopic + Debug + Send + Sync,
+        CmdWrp: Decode + BindTopic + Debug + Send + Sync,
+        CmdPlr: Peeler<Wrp = CmdWrp, Msg = PushCommand<Cmd>> + Send + Sync,
+        Event: Decode + BindTopic + Debug + Send + Sync,
+        EventWrp: Decode + BindTopic + Debug + Send + Sync,
+        EventPlr: Peeler<Wrp = EventWrp, Msg = Event> + Send + Sync,
+        QReq: Serialize + DeserializeOwned + Debug,
+        QResp: Serialize + DeserializeOwned + Debug,
+        Con: NativeContract<Cmd = Cmd, Event = Event, QReq = QReq, QResp = QResp> + Send + Sync,
+        PushCommand<Cmd>: Decode + BindTopic + Debug,
+    {
+        fn id(&self) -> ContractId {
+            self.contract.id()
+        }
+
+        fn handle_query(
+            &mut self,
+            origin: Option<&runtime::AccountId>,
+            req: OpaqueQuery,
+        ) -> Result<OpaqueReply, OpaqueError> {
+            serde_json::to_value(
+                self.contract.handle_query(
+                    origin,
+                    deopaque_query(req)
+                        .map_err(|_| error_msg("Malformed request"))?
+                        .request,
+                ),
+            )
+            .map_err(|_| error_msg("serde_json serilize failed"))
+        }
+
+        fn process_events(&mut self, env: &mut ExecuteEnv) {
+            let storage = env.storage;
+            let key_map =
+                |topic: &phala_mq::Path| storage.get(&storage_prefix_for_topic_pubkey(topic));
+            let osp_mq = OspMq::new(&self.ecdh_key, &self.send_mq, &key_map);
+            let context = NativeContext {
+                block_number: env.block_number,
+                mq: &self.send_mq,
+                osp_mq,
+            };
+            loop {
+                let ok = phala_mq::select! {
+                    next_cmd = self.cmd_rcv_mq => match next_cmd {
+                        Ok(Some((_, cmd, origin))) => {
+                            // TODO.kevin: allow all kind of origin to send commands to contract?
```

Comment:
Yes, but we can leave other origins (like Pallet) when we encounter the requirement in the future